src/sync/list.rs

```diff
@@ -212,37 +241,50 @@ impl<'g, T: 'g, C: Container<T>> Iterator for Iter<'g, T, C> {
                 // This entry was removed. Try unlinking it from the list.
                 let succ = succ.with_tag(0);
 
-                match self.pred.compare_and_set_weak(
+                if self.curr.tag() != 0 {
```

Comment:
After changing the behavior of the `Err` branch below (restarting from the beginning), it seems `self.cur.tag()` should always be zero.  If so, we can remove this branch.

Reply:
It seems you are right! Updated.